File: src/main/java/_sorting/InsertionSort.java
package _sorting;

import verifications.SortedVerification;

public class InsertionSort {

    private static Integer[] insertionSort(Integer[] input) {
        if (input == null || input.length <= 1) {
            return input;
        }
        //The array [0 to i-1] is already sorted
        //This loops stops when i = input.length
        //Therefore, by invariant [0 to input.length-1] is sorted
        for (int i = 1; i < input.length; i++) {
            Integer value = input[i];
            int j = i;
            while (j > 0 && input[j - 1] > value) {
                input[j] = input[j - 1];
                j--;
            }
            input[j] = value;

        }
        return input;
    }

    public static void main(String[] args) {
        SortedVerification.Companion.sortSmallArray(InsertionSort::insertionSort);
    }
}

File: src/main/kotlin/sorting/InsertionSort.kt
package sorting

import verifications.SortedVerification

class InsertionSort {
    companion object {
        fun insertionSort(input: Array<Int>): Array<Int> {
            //For each next element, we find the location where this one belongs.
            //This one iterates the sorted elements on the left of the array
            for (i in input.indices) {
                //Iterate backwards and swap the element
                val item = input[i]
                var j = i - 1
                while (j >= 0 && input[j] > item) {
                    input[j + 1] = input[j]
                    input[j] = item
                    j--
                }
            }
            return input
        }
    }
}

fun main() {
    SortedVerification.sortArray { InsertionSort.insertionSort(it) }
}